// @colsen1991/ngx-translate-extract-marker/esm2020/extras/ngx-translate-extract-marker.module.mjs
import { NgModule } from '@angular/core';
import { MarkerDirective } from './ngx-translate-extract-marker.directive';
import { MarkerPipe } from './ngx-translate-extract-marker.pipe';
import * as i0 from "@angular/core";
export class MarkerModule {
}
MarkerModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: MarkerModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
MarkerModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "14.3.0", ngImport: i0, type: MarkerModule, declarations: [MarkerDirective,
        MarkerPipe], exports: [MarkerDirective,
        MarkerPipe] });
MarkerModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: MarkerModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: MarkerModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [
                        MarkerDirective,
                        MarkerPipe
                    ],
                    imports: [],
                    exports: [
                        MarkerDirective,
                        MarkerPipe
                    ]
                }]
        }] });
