// @colsen1991/ngx-translate-extract-marker/esm2020/extras/ngx-translate-extract-marker.directive.mjs
import { Directive } from '@angular/core';
import * as i0 from "@angular/core";
export class MarkerDirective {
}
MarkerDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: MarkerDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive });
MarkerDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "14.3.0", type: MarkerDirective, selector: "[marker]", ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: MarkerDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[marker]',
                }]
        }] });
export class MarkerDirectiveStandalone {
}
MarkerDirectiveStandalone.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: MarkerDirectiveStandalone, deps: [], target: i0.ɵɵFactoryTarget.Directive });
MarkerDirectiveStandalone.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "14.3.0", type: MarkerDirectiveStandalone, isStandalone: true, selector: "[marker]", ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: MarkerDirectiveStandalone, decorators: [{
            type: Directive,
            args: [{
                    selector: '[marker]',
                    standalone: true,
                }]
        }] });
