// @colsen1991/ngx-translate-extract-marker/esm2020/extras/ngx-translate-extract-marker.pipe.mjs
import { Pipe } from '@angular/core';
import * as i0 from "@angular/core";
export class MarkerPipe {
    transform(key) {
        return key;
    }
}
MarkerPipe.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: MarkerPipe, deps: [], target: i0.ɵɵFactoryTarget.Pipe });
MarkerPipe.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "14.0.0", version: "14.3.0", ngImport: i0, type: MarkerPipe, name: "marker" });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: MarkerPipe, decorators: [{
            type: Pipe,
            args: [{
                    name: 'marker'
                }]
        }] });
export class MarkerPipeStandalone {
    transform(key) {
        return key;
    }
}
MarkerPipeStandalone.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: MarkerPipeStandalone, deps: [], target: i0.ɵɵFactoryTarget.Pipe });
MarkerPipeStandalone.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "14.0.0", version: "14.3.0", ngImport: i0, type: MarkerPipeStandalone, isStandalone: true, name: "marker" });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: MarkerPipeStandalone, decorators: [{
            type: Pipe,
            args: [{
                    name: 'marker',
                    standalone: true
                }]
        }] });
