// @colsen1991/ngx-translate-extract-marker/fesm2015/colsen1991-ngx-translate-extract-marker-extras.mjs
import * as i0 from '@angular/core';
import { Pipe, Directive, NgModule } from '@angular/core';

class MarkerPipe {
    transform(key) {
        return key;
    }
}
MarkerPipe.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: MarkerPipe, deps: [], target: i0.ɵɵFactoryTarget.Pipe });
MarkerPipe.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "14.0.0", version: "14.3.0", ngImport: i0, type: MarkerPipe, name: "marker" });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: MarkerPipe, decorators: [{
            type: Pipe,
            args: [{
                    name: 'marker'
                }]
        }] });
class MarkerPipeStandalone {
    transform(key) {
        return key;
    }
}
MarkerPipeStandalone.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: MarkerPipeStandalone, deps: [], target: i0.ɵɵFactoryTarget.Pipe });
MarkerPipeStandalone.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "14.0.0", version: "14.3.0", ngImport: i0, type: MarkerPipeStandalone, isStandalone: true, name: "marker" });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: MarkerPipeStandalone, decorators: [{
            type: Pipe,
            args: [{
                    name: 'marker',
                    standalone: true
                }]
        }] });

class MarkerDirective {
}
MarkerDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: MarkerDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive });
MarkerDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "14.3.0", type: MarkerDirective, selector: "[marker]", ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: MarkerDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[marker]',
                }]
        }] });
class MarkerDirectiveStandalone {
}
MarkerDirectiveStandalone.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: MarkerDirectiveStandalone, deps: [], target: i0.ɵɵFactoryTarget.Directive });
MarkerDirectiveStandalone.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "14.3.0", type: MarkerDirectiveStandalone, isStandalone: true, selector: "[marker]", ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: MarkerDirectiveStandalone, decorators: [{
            type: Directive,
            args: [{
                    selector: '[marker]',
                    standalone: true,
                }]
        }] });

class MarkerModule {
}
MarkerModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: MarkerModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
MarkerModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "14.3.0", ngImport: i0, type: MarkerModule, declarations: [MarkerDirective,
        MarkerPipe], exports: [MarkerDirective,
        MarkerPipe] });
MarkerModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: MarkerModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: MarkerModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [
                        MarkerDirective,
                        MarkerPipe
                    ],
                    imports: [],
                    exports: [
                        MarkerDirective,
                        MarkerPipe
                    ]
                }]
        }] });

/**
 * Generated bundle index. Do not edit.
 */

export { MarkerDirective, MarkerDirectiveStandalone, MarkerModule, MarkerPipe, MarkerPipeStandalone };
